pkg/velaql/providers/query/tree.go

```diff
@@ -0,0 +1,439 @@
+		if len(svc.Status.LoadBalancer.Ingress) > 0 {
+			health.Status = types.HealthStatusHealthy
+		} else {
+			health.Status = types.HealthStatusProgressing
+		}
+	}
+	return &health, nil
+}
+
+func checkResourceStatus(obj unstructured.Unstructured) (*types.HealthStatus, error) {
+	group := obj.GroupVersionKind().Group
+	kind := obj.GroupVersionKind().Kind
+	var checkFunc healthyCheckFunc
+	switch group {
+	case "":
+		switch kind {
+		case "Pod":
+			checkFunc = checkPodStatus
+		case "Service":
+			checkFunc = checkServiceStatus
+		case "PersistentVolumeClaim":
+			checkFunc = checkPVCHealthStatus
+		}
+	case "apps":
+		switch kind {
+		case "ReplicaSet":
+			checkFunc = checkReplicaSetStatus
+		default:
+		}
+	}
+	if checkFunc != nil {
+		return checkFunc(obj)
+	}
+	return &types.HealthStatus{Status: types.HealthStatusHealthy}, nil
+}
+
+func fetchObjectWithResourceTreeNode(ctx context.Context, cluster string, k8sClient client.Client, resource types.ResourceTreeNode) (*unstructured.Unstructured, error) {
+	o := unstructured.Unstructured{}
+	o.SetAPIVersion(resource.APIVersion)
+	o.SetKind(resource.Kind)
+	o.SetNamespace(resource.Namespace)
+	o.SetName(resource.Name)
+	err := k8sClient.Get(multicluster.ContextWithClusterName(ctx, cluster), types2.NamespacedName{Namespace: resource.Namespace, Name: resource.Name}, &o)
+	if err != nil {
+		return nil, err
+	}
+	return &o, nil
+}
+
+func listItemByRule(clusterCTX context.Context, k8sClient client.Client, resource ResourceType,
+	parentObject unstructured.Unstructured, specifiedFunc genListOptionFunc, defaultFunc genListOptionFunc) ([]unstructured.Unstructured, error) {
+
+	itemList := unstructured.UnstructuredList{}
+	itemList.SetAPIVersion(resource.APIVersion)
+	itemList.SetKind(fmt.Sprintf("%sList", resource.Kind))
+	var err error
+	if specifiedFunc == nil && defaultFunc == nil {
+		// if the relationShip between parent and child hasn't defined by any genListOption, list all subResource and filter by ownerReference UID
+		err = k8sClient.List(clusterCTX, &itemList)
+		if err != nil {
+			return nil, err
+		}
+		var res []unstructured.Unstructured
+		for _, item := range itemList.Items {
+			for _, reference := range item.GetOwnerReferences() {
+				if reference.UID == parentObject.GetUID() {
+					res = append(res, item)
+				}
+			}
+		}
+		return res, nil
+	}
+	var listOptions client.ListOptions
+	if specifiedFunc != nil {
+		//  specified func will override the default func
+		listOptions, err = specifiedFunc(parentObject)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		listOptions, err = defaultFunc(parentObject)
+		if err != nil {
+			return nil, err
+		}
+	}
+	err = k8sClient.List(clusterCTX, &itemList, &listOptions)
+	if err != nil {
+		return nil, err
+	}
+	return itemList.Items, nil
+}
+
+func iteratorChildResources(ctx context.Context, cluster string, k8sClient client.Client, parentResource types.ResourceTreeNode) ([]types.ResourceTreeNode, error) {
```

Comment:
we should add some limit for the depth, I'm afraid there will be potential performance issues.

We can add a default max-depth to 5 or 10, if you don't have any recommend value. You can check some typical scenarios and find one.

Reply:
In my tests, most scenario cannot reach 5, helmRelease -> deployment -> rs -> pod

Reply:
Then 5 is safe here.